TrackPropagation/Geant4e/test/Geant4ePropagatorAnalyzer.cc

```diff
@@ -331,18 +331,20 @@ void Geant4ePropagatorAnalyzer::analyze(const edm::Event &iEvent, const edm::Eve
   ///////////////////////////////////////
   // Get the sim tracks & vertices
   Handle<SimTrackContainer> simTracks;
-  iEvent.getByLabel<SimTrackContainer>(G4TrkSrc_, simTracks);
+  iEvent.getByLabel<SimTrackContainer>("g4SimHits", "", simTracks);
   if (!simTracks.isValid()) {
     LogWarning("Geant4e") << "No tracks found" << std::endl;
-    return;
+    //    return;
   }
   LogDebug("Geant4e") << "G4e -- Got simTracks of size " << simTracks->size();
+  std::cout << "Geant4e "
+            << "G4e -- Got simTracks of size " << simTracks->size() << std::endl;
 
   Handle<SimVertexContainer> simVertices;
-  iEvent.getByLabel<SimVertexContainer>(G4VtxSrc_, simVertices);
+  iEvent.getByLabel<SimVertexContainer>("g4SimHits", simVertices);
   if (!simVertices.isValid()) {
-    LogWarning("Geant4e") << "No tracks found" << std::endl;
-    return;
+    LogWarning("Geant4e") << "No verticess found" << std::endl;
```

Comment:
typo.

Reply:
thanks for spotting.
Once we find how to get the products will fix this as well

Reply:
Probably unrelated, but about

>the test does not find products (getByLabel broken?)

Shouldn't be tokens used instead? https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideEDMGetDataFromEvent#get_ByToken

AFAIK `getByLabel` is deprecated, but I would have expected a different error message.

Reply:
> AFAIK `getByLabel` is deprecated, but I would have expected a different error message.

Correct, and I would have expected too to see a different exception (complaining about missing `consumes()` calls).

Reply:
deprecated  should not mean not working... (or message should indeed say something different)

Reply:
Does one really has to consume each single product used?
and create tokens etc just for a single very stupid analyzer?
One cannot just say I consume everything in the input file?
I understand products to be produced, but for those in the input file is a bit abusive to ask all those boilerplate....

is there a script that generate the boilerplate?
I will not write it by hand!

Reply:
> Does one really has to consume each single product used? 

Yes.

> and create tokens etc just for a single very stupid analyzer?

Yes.

> One cannot just say I consume everything in the input file? 

No.

> I understand products to be produced, but for those in the input file is a bit abusive to ask all those boilerplate....

The consumes information is used also for input file to read in only those branches that are actually needed by the module.

> is there a script that generate the boilerplate?

I'm not aware of any.

Reply:
I've added
```
@@ -147,6 +147,9 @@ Geant4ePropagatorAnalyzer::Geant4ePropagatorAnalyzer(const edm::ParameterSet &iC
       thePropagator(nullptr),
       G4VtxSrc_(iConfig.getParameter<edm::InputTag>("G4VtxSrc")),
       G4TrkSrc_(iConfig.getParameter<edm::InputTag>("G4TrkSrc")) {
+
+      consumes(G4TrkSrc_);
+
````

in the constructor.

getByLabel still produces the very same error as above

Reply:
Just to note that the exception message has been improved for this particular corner case in https://github.com/cms-sw/cmssw/pull/38875.

About
> ```
> @@ -147,6 +147,9 @@ Geant4ePropagatorAnalyzer::Geant4ePropagatorAnalyzer(const edm::ParameterSet &iC
>        thePropagator(nullptr),
>        G4VtxSrc_(iConfig.getParameter<edm::InputTag>("G4VtxSrc")),
>        G4TrkSrc_(iConfig.getParameter<edm::InputTag>("G4TrkSrc")) {
> +
> +      consumes(G4TrkSrc_);
> +
> ```

in this case the `consumes()` call would need an explicit type (because it can't infer it from anywhere) along
> ```
> @@ -147,6 +147,9 @@ Geant4ePropagatorAnalyzer::Geant4ePropagatorAnalyzer(const edm::ParameterSet &iC
>        thePropagator(nullptr),
>        G4VtxSrc_(iConfig.getParameter<edm::InputTag>("G4VtxSrc")),
>        G4TrkSrc_(iConfig.getParameter<edm::InputTag>("G4TrkSrc")) {
> +
> +      consumes<std::vector<SimTrack>>(G4TrkSrc_);
> +
> ```

In that case the `getByLabel()` should still work (even if `getByToken()` is strongly favored). The `consumes()` call itself has been required since 2016.